### PR TITLE
feat(bundle): optionally displays labels associated to a bundle

### LINKED ISSUE #235: Datamon bundle list should include labels
Right now datamon bundle list returns something like this
```
/tmp/datamon bundle list --repo flood-postgres
Using config file: /home/developer/.datamon/datamon.yaml
1N9KQjinEjRGtKovxksVG6biS3c , 2019-06-25 18:37:37.934833707 +0000 UTC , First version of flood pg data
1N9jaJkPxbnZyPlz5HL16N3C5xg , 2019-06-25 22:04:28.307925983 +0000 UTC , Greatly prune flood PG data to reduce backup time
1NCMOUOqdxqkrXMCIkqmZXbjIIZ , 2019-06-26 20:23:12.978399405 +0000 UTC , Backup from deployed data
1OCQ2W2lcIoRWQnBvk8r5D20J6B , 2019-07-18 19:41:31.151859934 +0000 UTC , Update flood alert thresholds
1OCf7RUcyfgecxrMUcwuko8Loex , 2019-07-18 21:45:31.68074457 +0000 UTC , Update flood alert thresholds
1OCg1ZgwBgdx0Oz3K1gjdGAqRuq , 2019-07-18 21:52:57.59154874 +0000 UTC , Update flood alert thresholds
```

I would like to see it return something like this:

````
/tmp/datamon bundle list --repo flood-postgres
Using config file: /home/developer/.datamon/datamon.yaml
1N9KQjinEjRGtKovxksVG6biS3c , v1.0.0, 2019-06-25 18:37:37.934833707 +0000 UTC , First version of flood pg data
1N9jaJkPxbnZyPlz5HL16N3C5xg , v1.0.1, 2019-06-25 22:04:28.307925983 +0000 UTC , Greatly prune flood PG data to reduce backup time
1NCMOUOqdxqkrXMCIkqmZXbjIIZ , <no label>, 2019-06-26 20:23:12.978399405 +0000 UTC , Backup from deployed data
1OCQ2W2lcIoRWQnBvk8r5D20J6B , <no label>, 2019-07-18 19:41:31.151859934 +0000 UTC , Update flood alert thresholds
1OCf7RUcyfgecxrMUcwuko8Loex , <no label>, 2019-07-18 21:45:31.68074457 +0000 UTC , Update flood alert thresholds
1OCg1ZgwBgdx0Oz3K1gjdGAqRuq , v1.0.2, 2019-07-18 21:52:57.59154874 +0000 UTC , Update flood alert thresholds
```

### PR DESCRIPTION
**Pull request by fredbi**
* fixes #235

Signed-off-by: Frederic BIDON <frederic@oneconcern.com>

## Note to reviewers

This I consider an early experiment to gauge competing ideas about how to deal with CLI features that pertain to the "porcelain" layer.

While I am all in for some DSL approach that would eventually result in neater code, I have difficulties figuring out how it would actually look like.

This PR plays here the devil's advocate and is my 2 cents contribution to this interesting debate about implementing porcelain, with some code-grounded facts. No big deal if this PR is discarded....

I wanted to figure out a more accurate idea about how much of code/complexity such user requirements actually entail when implemented with the basic tooling the CLI is currently providing.

Regarding expected results, I followed the approach recommended by @ransomw1c in #235 (";" separators) and did not try to over-engineer it with some asynchronous fetching of labels.

### CODE REVIEW

File: cmd/datamon/cmd/bundle_get.go [hunk 1]
```diff
@@ -57,8 +58,24 @@ exits with ENOENT status otherwise.`,
...
+			data := struct {
+				model.BundleDescriptor
+				Labels string
+			}{
+				BundleDescriptor: bundle.BundleDescriptor,
+			}
+			data.Labels = displayBundleLabels(bundle.BundleDescriptor.ID, labels)
+			err = bundleDescriptorTemplate(datamonFlags.bundle.WithLabels).Execute(&buf, data)
+		} else {
+			err = bundleDescriptorTemplate(datamonFlags.bundle.WithLabels).Execute(&buf, bundle.BundleDescriptor)
```
Comment (ransomw1c) [inline cmd/datamon/cmd/bundle_get.go]: `model.BundleDescriptor` doesn't have a `Labels` `reflect.StructField`, so if `datamonFlags.bundle.WithLabels` is true and `len(labels)` is zero (i.e. if the `--with-labels` flag is passed and the bundle in question happens not to have any labels, this this will pass `true` to `bundleDescriptorTemplate` and we'll be attempting to call this `Execute` with a template string

```
{{.ID}} ,{{.Labels}}, {{.Timestamp}} , {{.Message}}
```

and a `model.BundleDescriptor` `interface{}` parameter, which will result in an error?

Comment (fredbi) [reply cmd/datamon/cmd/bundle_get.go]: @ransomw1c ok will do.

**Comment by ransomw1c** [comment]
just reading the note to reviewers.

the thing about calling this a "porcelain layer" change is that it's actually using some additional "plumbing," `core.ListLabels` that wasn't previously consumed by the `cmd/` package.

if this deduped and re-used the `list labels` functionality such that the `bundle list` and `list labels` relied on literally the exact same data path to retrieve labels, then i think it would be clearer why the requirement could be implemented independent of adding the additional `getLabels` "plumbing" to `pkg/`.

however, as i say, i don't think this a pure porcelain change.  it's like if we were shipping `async def someFunc()`  in some alternative universe where the cli was implemented via python bindings, then started shipping `def someFuncSync()`.  even if the data paths of `someFunc` and `someFuncSync` are pretty much the same under the hood, that's not an api guarantee.  from downstream, they're separate functions.  or at least the water is murkier than it would be in the situation described in the previous paragraph.

**Comment by fredbi** [comment]
**Re remark on python bindings**

@ransomw1c on the slack channel I posted a proposal about figuring out some python bindings (I've done that in the past, binding python 2 and 3 to a go lib, so I could do this again). This could really be a nice thing, but I didn't receive any comment on that post.

File: cmd/datamon/cmd/bundle_get.go (hunk 1 above)
Comment (fredbi) [reply cmd/datamon/cmd/bundle_get.go]: @ransomw1c there is no failure when no label is set on the bundle. I am now adding a test to prove that.
TL,DR:
```
./datamon bundle get --repo fred-test-repo --with-labels
Using bundle: 1WAKWqimd3kKt24oI3qg3XGzXsu
{"level":"info","ts":1578641986.0438097,"caller":"core/bundle_unpack.go:258","msg":"kicking off filelist download","concurrent Filelist Downloads":10}
{"level":"info","ts":1578641986.044021,"caller":"core/bundle_unpack.go:277","msg":"preallocating bundle entries","max entries":0}
1WAKWqimd3kKt24oI3qg3XGzXsu ,<no label>, 2020-01-09 17:21:54.772393594 +0100 CET , xyz
```

```
./datamon label set --bundle 1WAKWqimd3kKt24oI3qg3XGzXsu  --repo fred-test-repo --label my-label
./datamon bundle get --repo fred-test-repo --with-labels
Using bundle: 1WAKWqimd3kKt24oI3qg3XGzXsu
{"level":"info","ts":1578642033.7185533,"caller":"core/bundle_unpack.go:258","msg":"kicking off filelist download","concurrent Filelist Downloads":10}
{"level":"info","ts":1578642033.7187045,"caller":"core/bundle_unpack.go:277","msg":"preallocating bundle entries","max entries":0}
1WAKWqimd3kKt24oI3qg3XGzXsu ,(my-label), 2020-01-09 17:21:54.772393594 +0100 CET , xyz
```

Comment (fredbi) [reply cmd/datamon/cmd/bundle_get.go]:
```
./datamon bundle list --repo fred-test-repo --with-labels
1VqLCB9upklFMAr3kch5Xhv2FXv ,<no label>, 2020-01-02 15:32:27.948443111 +0100 CET , upload a bundle
1VqM633GnZ8Mh1Dt38rXWlcDKHD ,<no label>, 2020-01-02 15:39:21.299499846 +0100 CET , Mutable mount
1VqMMh2XqjM9JZVRvsK5BJfcf5b ,<no label>, 2020-01-02 15:41:22.454642726 +0100 CET , Mutable mount
1VqMZtZzCRtoWSQIgarW9EBaoEl ,(mutable-mount-fred), 2020-01-02 15:42:28.202893153 +0100 CET , Mutable mount
1VqMoiNzR2ZXaaoQfzhXj6JXsP7 ,(mutable-mount-fred-4), 2020-01-02 15:45:03.56740456 +0100 CET , Mutable mount 4
1VqNBo70inzbiRTPkleMrl1mhVR ,<no label>, 2020-01-02 15:48:50.890137366 +0100 CET , upload a bundle again
1VqRkcyVagyL8ssEfo1uCI7pB3i ,(mutable-mount-fred-5), 2020-01-02 15:52:37.756544005 +0100 CET , Mutable mount 5
1VqRwfofF0To4d9OY5VgAAxrVxN ,<no label>, 2020-01-02 16:27:57.498375401 +0100 CET , upload a bundle yet again
1VqS4BwCRSIFSFQDAmqc2K4NMdv ,(mutable-mount-fred-6), 2020-01-02 16:28:55.944993424 +0100 CET , Mutable mount 6
1VqSA0iNLRIr6L3TpuWNloFJRov ,(un label), 2020-01-02 16:29:44.555730364 +0100 CET , upload a bundle yet again
1VqSirBWrUDC9Pa0KQIiJp64sUW ,(mutable-mount-fred-7), 2020-01-02 16:30:43.598354541 +0100 CET , Mutable mount 7
1WAKWqimd3kKt24oI3qg3XGzXsu ,(my-label), 2020-01-09 17:21:54.772393594 +0100 CET , xyz
```

Comment (fredbi) [reply cmd/datamon/cmd/bundle_get.go]: @ransomw1c corner case of repo without label fixed, CLI test added to cover that case

**Comment by ransomw1c** [comment]
quick note on collective opinions here:  after discussion with Joao last week, i suppose it's better understood that my remarks about shipping a Python api to access Datamon functionality in addition to the cli api are purely motivated by attempting to fulfill feature requests related to the cli api without turning `cmd/datamon/cmd` into spaghetti, and of course we'll all agree that we don't want spaghetti.

at the same time, there are ways to keep `cmd/datamon/cmd` maintainable without negotiating cli feature requests into a Python api – refactoring cobra use or using a different cli lib altogether to name a couple.

supposing tentative plans for the couple weeks (see Jira RES-1809) occur, one possible scenario – a likely one, even, imo – is that we'll continue to cruise toward cutting v2 with a mild amount of tech debt in `datamon/cmd` and meanwhile condense `hack/fuse-demo/wrap_datamon.sh` (note that this prototype we're about to ship to production is still in `fuse-demo/`, btw) into a (hopefully much terser) Python script.

so i'll propose to continue worrying about the maintainability of `cmd/datamon/cmd`, except only to first try to make it more maintainable using golang only.  meanwhile, there will be some Pythonic faceting of the binary for a totally different raison, the maintainability of the sidecar.  and, imo, writing the Pythonic faceting for that purpose could also help get clarity on using python as a means to make the desktop CLI more maintainable.

**Comment by kerneltime** [comment]
We can park the branch for now. We need to think more about this feature.

**Comment by fredbi** [comment]
@kerneltime sure. I wanted to close it to. Ransom did rightfully point out that this should go as porcelain. Organizing the porcelain layer is going to be the big code shuffle that will come together with the python SDK.
